### PR TITLE
fix(component): persist selection across TableNext pages

### PR DESCRIPTION
**Pull request by chanceaclark**
## What?

Enables `TableNext` to persist selected items across pagination.

## Why?

For each page, we were returning the indexes of items without accounting for pagination. If you were on page 2, the indexes you were getting back were `{ 0: true, 1: true, 2.0: true, ...etc }`. This made it difficult to both Selected all products across pages and persist selected items across pages. This update aligns with our vision of moving toward using `@tanstack/react-table` or other similar headless table API's to power our `TableNext` component.

## Screenshots/Screen Recordings

https://user-images.githubusercontent.com/10539418/192371874-5c04a746-750c-4a56-8694-293a6e98134f.mov

## Testing/Proof

Here's the `dev.tsx` page I used if you want to test it out yourself:
```tsx
import { TableNext } from '@bigcommerce/big-design/dist/es/components/TableNext';
import React, { useEffect, useState } from 'react';

interface Item {
  name: string;
  children?: Item[];
}

const items = new Array(25).fill(undefined).map<Item>((_, index) => ({
  name: `Hello ${index}`,
  children: [{ name: `Child ${index}` }, { name: `Child ${index}` }],
}));

const DevPage = () => {
  const [currentPage, onPageChange] = useState(1);
  const [itemsPerPage, onItemsPerPageChange] = useState(10);
  const [selectedItems, onSelectionChange] = useState<Record<string, true>>({});
  const [pagedItems, setPagedItems] = useState<typeof items>([]);
  const [expandedRows, onExpandedChange] = useState({});

  useEffect(() => {
    setPagedItems(
      items.filter(
        (_item, index) =>
          index >= (currentPage - 1) * itemsPerPage && index < currentPage * itemsPerPage,
      ),
    );
  }, [currentPage, itemsPerPage]);

  return (
    <TableNext
      columns={[{ hash: 'name', header: 'Name', render: ({ name }) => name }]}
      expandable={{
        expandedRows,
        onExpandedChange,
        expandedRowSelector: ({ children }) => {
          if (children) {
            return children;
          }

          return undefined;
        },
      }}
      itemName="Worlds"
      items={pagedItems}
      pagination={{
        currentPage,
        onPageChange,
        itemsPerPage,
        totalItems: items.length,
        onItemsPerPageChange,
        itemsPerPageOptions: [10, 20, 30],
      }}
      selectable={{ onSelectionChange, selectedItems }}
    />
  );
};

export default DevPage;
```

```ts
// global.d.ts

// ...

declare module '@bigcommerce/big-design/dist/es/components/TableNext' {
  export * from '@bigcommerce/big-design/dist/components/TableNext';
}

### CODE REVIEW

File: packages/big-design/src/components/TableNext/SelectAll/helpers.ts [hunk 1]
```diff
@@ -1,23 +1,12 @@
...
-) {
-  return items.reduce((acc, _parentRow, parentRowIndex) => {
-    if (selectedItems[parentRowIndex] !== undefined) {
-      return acc + 1;
-    }
+type SelectAllRowsArg<T> = Omit<SelectAllProps<T>, 'onChange'>;
 
-    return acc;
-  }, 0);
+export function getTotalSelectedItems(selectedItems: TableSelectable['selectedItems']) {
```
Comment (bc-juanvasquez) [inline packages/big-design/src/components/TableNext/SelectAll/helpers.ts:8]: looks like this changing from getting 'total parents' to 'total parents and children' ? Do we want to include children in total count?

**Comment by jorgemoya** [comment]
Maybe this was discussed before, but shouldn't select 25 select them all? I would've expected all items to be selected across pagination. This still works to persist selection.

Is this just a different ticket issue?

File: packages/big-design/src/components/TableNext/SelectAll/helpers.ts (hunk 1 above)
Comment (chanceaclark) [reply packages/big-design/src/components/TableNext/SelectAll/helpers.ts:8]: Ah you're right, good catch!

**Comment by chanceaclark** [comment]
> Maybe this was discussed before, but shouldn't select 25 select them all? I would've expected all items to be selected across pagination. This still works to persist selection.
> 
> Is this just a different ticket issue?

Yeah we've discussed this before. The "Select all" checkbox should just select all the items on the current page. We have another pattern to "Select everything" using bulk actions.

**Comment by bc-juanvasquez** [comment]
Screen Shot 2022-09-26 at 4 27 27 PM

I believe the 'Selected Worlds' label should only reflect selected parents. At least that's whats reflected in the implementation mocks..

File: packages/big-design/src/components/TableNext/SelectAll/helpers.ts [hunk 2]
```diff
@@ -60,65 +51,101 @@ export function areSomeInPageSelected<T>({
...
-}: Omit<SelectAllRowsArg<T>, 'selectedItems'>): TableSelectable['selectedItems'] {
+function deselectAllOnCurrentPage<T>(params: SelectAllRowsArg<T>) {
+  const { items, selectedItems, pagination } = params;
+
+  const filteredSelectedItems = Object.keys(selectedItems)
+    .filter((selectedKey) => {
+      const [parentIndex] = selectedKey.split('.').map((key) => parseInt(key, 10));
+      const item = items.find((_, index) => getPagedIndex(index, pagination) === parentIndex);
+
+      if (item) {
```
Comment (bc-juanvasquez) [inline packages/big-design/src/components/TableNext/SelectAll/helpers.ts:111]: return Boolean(item) ?

**Review by bc-juanvasquez — APPROVED** [review]

**Review by jorgemoya — APPROVED** [review]
Looks good to me.

**Pull request merged**